compiler-rt/lib/asan/asan_globals.cpp

```diff
@@ -147,14 +163,21 @@ static void CheckODRViolationViaIndicator(const Global *g) {
     *odr_indicator = REGISTERED;
     return;
   }
+
+  // Fetch globals with the same ODR indicator.
+  auto *relevant_globals_lookup =
+      map_of_globals_by_indicator.find(g->odr_indicator);
+  if (!relevant_globals_lookup)
+    return;
+
+  ListOfGlobals *relevant_globals = relevant_globals_lookup->second;
   // If *odr_indicator is DEFINED, some module have already registered
   // externally visible symbol with the same name. This is an ODR violation.
-  for (ListOfGlobals *l = list_of_all_globals; l; l = l->next) {
-    if (g->odr_indicator == l->g->odr_indicator &&
```

Comment:
the patch does not take into account that global can be unregistered.
As is it will stay in map with invalid indicator.

Reply:
Good point! This is in line with the current behavior where it also stays in the list with an `UNREGISTERED` indicator. If we're ok keeping just the first global in the map, then removing from the map when unregistering is easy. If we need the whole list, removing is also not hard and should be still fast enough.

LMK if it's important to have removal from the map as part of this patch (or maybe as a patch on top).

Reply:
Instead of removal checking that indicator matches is enough

we can't keep just first one with UNREGISTERED, if we unregister it we don't know if there are other with same indicator are still registered

Reply:
> Instead of removal checking that indicators still match is enough

@vitalybuka could you expand on what you mean by 'indicators still match'?

Looking at the code for `UnregisterGlobal`, we update the value of whatever the indicator points to to `UNREGISTERED` (which affects all instances that share the same indicator) but the indicator address is unchanged.
https://github.com/llvm/llvm-project/blob/99b2444c7742b2c3a955660c75a95e063c03fc74/compiler-rt/lib/asan/asan_globals.cpp#L283-L284

I didn't find any tests around 'unregistering' so I added one in https://github.com/llvm/llvm-project/pull/100923/commits/01847d1dd72fb078d26f38973391f37d2815ec3c to demonstrate the behavior (it's the same on main and in this branch).

Reply:
LGTM
I was confused by `odr_indicator` vs `*odr_indicator`.